src/ui/public/state_management/state_storage/_tests__/hashed_item_store.js

```diff
@@ -0,0 +1,285 @@
+      });
+
+      it('returns true', () => {
+        expect(result).to.equal(true);
+      });
+    });
+
+    describe(`if the item doesn't exist in sessionStorage`, () => {
+      describe(`if there's storage space`, () => {
+        let sessionStorage;
+        let hashedItemStore;
+        const hash = 'a';
+        const item = JSON.stringify({});
+
+        beforeEach(() => {
+          sessionStorage = new StubBrowserStorage();
+          hashedItemStore = new HashedItemStore(sessionStorage);
+        });
+
+        it('updates the indexed item as recently touched', done => {
+          const time = Date.now();
+          setTimeout(() => {
+            try {
+              hashedItemStore.setItem(hash, item);
+              const touched = hashedItemStore._itemIndex.a.touched;
+              expect(touched).to.be.above(time);
+            } catch (e) {
+              return done(e);
+            }
+            done();
+          }, 1);
+        });
+
+        it('persists the item in sessionStorage', () => {
+          hashedItemStore.setItem(hash, item);
+          expect(sessionStorage.getItem(hash)).to.equal(item);
+        });
+
+        it('returns true', () => {
+          const result = hashedItemStore.setItem(hash, item);
+          expect(result).to.equal(true);
+        });
+      });
+
+      describe(`if there isn't storage space`, () => {
+        let sessionStorage;
+        let hashedItemStore;
+        let storageSizeLimit;
+        const hash = 'a';
+        const item = JSON.stringify({});
+
+        function calculateItemSize(itemCount) {
+          const itemSize = hash.length + item.length;
+          return itemSize * itemCount;
+        }
+
+        function calculateIndexSize(itemCount) {
+          const index = {};
+          for (var i = 0; i < itemCount; i++) {
+            const hash = i.toString();
+            index[hash] = {
+              hash,
+              touched: Date.now(),
+            };
+          }
+          return JSON.stringify(index).length;
+        }
+
+        beforeEach(done => {
+          sessionStorage = new StubBrowserStorage();
+
+          // Anticipate filling out storage with a persisted index (containing 2 items)
+          // and the 2 items themselves.
+          // NOTE: This is a very brittle test because it depends so heavily on the
+          // implementation of the tested code.
+          const itemCount = 2;
+          const itemsSize = calculateItemSize(itemCount);
+          const persistedIndexKeySize = HashedItemStore.PERSISTED_INDEX_KEY.length;
+          const persistedIndexValueSize = calculateIndexSize(itemCount);
+
+          storageSizeLimit = itemsSize + persistedIndexKeySize + persistedIndexValueSize;
+          sessionStorage._setSizeLimit(storageSizeLimit);
+
+          hashedItemStore = new HashedItemStore(sessionStorage);
+
+          // Add some items that will be removed.
+          hashedItemStore.setItem('1', item);
+          setTimeout(() => {
```

Comment:
you can use sinon's clock stubbing to do this synchronously, but this works fine too